core/src/main/java/com/spotify/metrics/core/ReservoirWithTtl.java

```diff
@@ -133,8 +133,8 @@ public int size() {
 
     @Override
     public void update(final long value) {
-        valueBuffer.add(new ValueAndTimestamp(value, now.get()));
-        valueBuffer.remove();
+        final int index = position.getAndIncrement() % bufferSize;
+        valueBuffer.set(index, new ValueAndTimestamp(value, now.get()));
```

Comment:
this will eventually throw a ArrayIndexOutOfBoundsException when index becomes <0 because position wrapped

Reply:
fixed by the bitwise and (& 0x7FFFFFFF)